sw/host/opentitanlib/src/transport/ti50emulator/emu.rs

```diff
@@ -2,36 +2,391 @@
+        }
+        Err(EmuError::StartFailureCause(format!(
+            "Ti50 sub-process spawn failure to many tries MAX_RETRY: {}",
+            MAX_RETRY
+        ))
+        .into())
+    }
+
+    /// The function tries to safely terminate the Emulator sub-process.
+    /// If the sub-process does not finish its execution after time
+    /// described in `TIMEOUT` * `MAX_RETRY`, use SIGKILL to force sub-process termination.
+    /// If all method fail, it returns an EmuError.
+    fn stop_process(&mut self) -> Result<()> {
+        if let Some(handle) = &self.proc {
+            let pid = handle.id() as i32;
+            signal::kill(Pid::from_raw(pid), Signal::SIGTERM)
+                .context("Stop sub-process using SIGTERM")?;
+            for _retry in 0..MAX_RETRY {
+                std::thread::sleep(TIMEOUT);
+                match signal::kill(Pid::from_raw(pid), None) {
+                    Ok(()) => {}
+                    Err(nix::Error::Sys(nix::errno::Errno::ESRCH)) => {
+                        self.state = EmuState::Off;
+                        self.proc = None;
+                        return Ok(());
+                    }
+                    Err(e) => {
+                        self.state = EmuState::Error;
+                        return Err(EmuError::StopFailureCause(format!(
+                            "Unexpected error querying process presence: {}",
+                            e
+                        ))
+                        .into());
+                    }
+                }
+            }
+            signal::kill(Pid::from_raw(pid), Signal::SIGKILL)
+                .context("Stop sub-process using SIGKILL")?;
+            std::thread::sleep(TIMEOUT);
+            match signal::kill(Pid::from_raw(pid), None) {
+                Err(nix::Error::Sys(nix::errno::Errno::ESRCH)) => {
+                    self.proc = None;
+                    self.state = EmuState::Off;
+                    return Ok(());
+                }
+                _ => {
+                    self.proc = None;
+                    self.state = EmuState::Error;
+                    return Err(EmuError::StopFailureCause(format!(
+                        "Unable to stop process pid:{}",
+                        pid
+                    ))
+                    .into());
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Method reset all internal states of Emulator to its default values.
+    fn reset_state(&mut self) -> Result<()> {
+        fs::remove_dir_all(&self.runtime_directory)?;
+        fs::create_dir(&self.runtime_directory)?;
+        self.current_args.clear();
+        self.current_args.insert(
+            String::from("exec"),
+            EmuValue::String(self.executable.clone()),
+        );
+        Ok(())
+    }
+
+    /// Update content of resource with data from `source`, overwrite file if it already exist.
+    fn update_resource(&self, key: &String, source: &PathBuf) -> Result<()> {
```

Comment:
Since you're not mutating `key`, you should use `&str`.
Since you're not mutating `source` here, its type should be `&Path` (&Path is to PathBuf as &str is to String).